Articles/Blog/2024-08-TechnicalPropertiesOfSustainableSoftware.md

```diff
@@ -0,0 +1,189 @@
+
+* **Weak Definition**: The same team of developers that originally created the software can continue to add new features and fix bugs in the software at a reasonable cost (i.e., where starting from scratch would be more expensive, even in the long term).
+
+* **Medium Definition**: New developers can reasonably make contributions to the software but these contributions are integrated back by the original development team.
+
+* **Strong Definition**: A different set of developers from the original developers can take over new feature development and bug fixes, and the original developers can go away (and the new team can do so cheaper than starting over from scratch).
+
+The ultimate instance of the "Strong Definition" is  the **Extreme Use Case for Sustainable Software**:
+Your project uses some external software Package X in such a way that it would be very difficult and expensive to rip it out and/or change the code to use something else.
+At some point, the developers of Package X go away, and no one is left in the development community to help extend or support Package X.
+What technical properties of Package X would make it so that you could continue support (with reasonable cost and risk) your own project's usage of Package X, which includes activities like adding some new features, porting to new platforms and compilers, and fixing bugs (i.e., that may be exposed by adding new features and porting to new platforms and upgrading upstream dependencies)?
+
+
+### Key Technical Properties of Sustainable Software
+
+The following technical properties have been widely recognized to lead to software that less expensive to extend and maintain (and is therefore more sustainable).
+
+**Open Source License**:
+Allow the user to change and use the software in any way that is needed for the success of their projects.
+Note that this does not require a fully open-source license in many cases.
+The license just has to allow your project to change the source and use the software as needed for your project.
+
+**Documented history of development**:
+The development history of the software (which can be captured in the version-control history and linked issue trackers) can answer many questions that are important to the future of the software.
+What factors caused the software to be in its current state?
+What requirements went into the development of the software? 
+(Might some of those requirements and features no longer be necessary for future versions of the software which would allow breaking some backward compatibility?)
+
+**Core domain model distillation document**:
+Has a clear documented core domain model (which is represented in the code or a clear documented mapping of the domain model to software) been given?<sup>[5]</sup>
+This document and model is needed to constrain the scope of the software and to continue to improve software cohesion and internal consistency with future changes.
+
+**Portable, well-documented, robust tests**:
```

Comment:
Might code formatting tools and static analyzers be included here?

Reply:
Code formatting tools (like clang-format) and static analyzers would go under the section "Some Key Practices for Creating and Maintaining Sustainable Software" under "Clean Understandable Code Practices" and "Testing Practices", respectively.  I will add those.

Reply:
Addressed in commit d21f5b6a30f5e776a2a474719ecff4b4f7622525?